File: 0087-scramble-string/0087-scramble-string.rb
# @param {String} s1
# @param {String} s2
# @return {Boolean}
def dfs(s1, s2, dp)
  key = [s1, s2]
  return dp[key] if dp.has_key?(key)
  if s1.length == 1
    ans = s1 == s2
    dp[key] = ans
    return ans
  end
  tally1 = [0]*26
  tally2 = [0]*26
  (0...s1.length).each do |i|
    tally1[s1[i].ord-'a'.ord] += 1
    tally2[s2[i].ord-'a'.ord] += 1
  end
  if tally1 != tally2
    dp[key] = false
    return false
  end
  
  ans = false
  (1...s1.length).each do |split|
    ans ||= dfs(s1[0...split], s2[0...split], dp) && \
      dfs(s1[split..-1], s2[split..-1], dp)
    ans ||= dfs(s1[0...split], s2[-split..-1], dp) && \
      dfs(s1[split..-1], s2[0...-split], dp)
  end
  dp[key] = ans
  ans
end

def is_scramble(s1, s2)
  dp = Hash.new
  dfs(s1, s2, dp)
end

File: 0087-scramble-string/0087-scramble-string.go
func isScramble(s1 string, s2 string) bool {
  dp := make(map[string]bool)
  return dfs(s1, s2, &dp)
}

func dfs(s1 string, s2 string, dp *map[string]bool) bool {
  key, ans := s1 + "-" + s2, false
  if v, ok := (*dp)[key]; ok {
    return v
  }
  if len(s1) == 1 {
    ans = s1 == s2
    (*dp)[key] = ans
    return ans
  }
  tally1, tally2 := make([]int, 26), make([]int, 26)
  for i := 0; i < 26; i++ {
    tally1[i] = 0
    tally2[i] = 0
  }
  for i := 0; i < len(s1); i++ {
    tally1[int(s1[i] - 'a')]++
    tally2[int(s2[i] - 'a')]++
  }
  if !slices.Equal(tally1, tally2) {
    (*dp)[key] = false
    return false
  }

  ans = false
  for split := 1; split < len(s1); split++ {
    ans = ans || dfs(s1[:split], s2[:split], dp) && dfs(s1[split:], s2[split:], dp)
    ans = ans || dfs(s1[:split], s2[len(s2) - split:], dp) && dfs(s1[split:], s2[:len(s2) - split], dp)
  }
  (*dp)[key] = ans
  return ans
}